File: src/tasks/cli/command/reword/RewordHandler.java
package tasks.cli.command.reword;

import static java.util.Objects.requireNonNull;

import io.reactivex.Single;
import omnia.cli.out.Output;
import omnia.data.cache.Memoized;
import tasks.cli.handlers.ArgumentHandler;
import tasks.cli.handlers.HandlerException;
import tasks.model.Task;
import tasks.model.TaskStore;

public final class RewordHandler implements ArgumentHandler<RewordArguments> {
  private final Memoized<TaskStore> taskStore;

  public RewordHandler(Memoized<TaskStore> taskStore) {
    this.taskStore = requireNonNull(taskStore);
  }

  @Override
  public Single<Output> handle(RewordArguments arguments) {
    String description = arguments.description().trim();
    if (description.isEmpty()) {
      throw new HandlerException("description cannot be empty or whitespace only");
    }

    return Single.fromCallable(taskStore::value)
        .flatMapCompletable(
            store ->
                store.mutateTask(
                    arguments.targetTask(),
                    mutator -> mutator.setLabel(arguments.description())))
        .andThen(taskStore.value().writeToDisk())
        .andThen(Single.just(arguments.targetTask()))
        .map(Task::render)
        .map(
            taskOutput ->
                Output.builder().append("Updated description: ").append(taskOutput).build());
  }
}

File: cli/src/me/strangepan/tasks/cli/feature/reword/RewordHandler.kt
package me.strangepan.tasks.cli.feature.reword

import io.reactivex.rxjava3.core.Single
import omnia.cli.out.Output
import omnia.cli.out.Output.Companion.builder
import omnia.data.cache.Memoized
import me.strangepan.tasks.cli.command.common.CommonArguments
import me.strangepan.tasks.cli.handler.ArgumentHandler
import me.strangepan.tasks.cli.handler.HandlerException
import me.strangepan.tasks.cli.model.render
import me.strangepan.tasks.engine.model.ObservableTaskStore

class RewordHandler(private val taskStore: Memoized<out ObservableTaskStore>) : ArgumentHandler<RewordArguments> {

  override fun handle(arguments: CommonArguments<out RewordArguments>): Single<Output> {
    val description = arguments.specificArguments().description().trim { it <= ' ' }
    if (description.isEmpty()) {
      throw HandlerException("description cannot be empty or whitespace only")
    }
    return Single.fromCallable { taskStore.value() }
        .flatMap { store ->
          store
              .mutateTask(arguments.specificArguments().targetTask()) {
                it.setLabel(arguments.specificArguments().description())
              }
              .map { it.third() }
        }
        .map { it.render() }
        .map { taskOutput ->
          builder()
              .append("Updated description: ")
              .append(taskOutput)
              .appendLine()
              .build()
        }
  }

}